wptrunner/executors/pytestrunner/fixtures.py

```diff
@@ -45,14 +46,79 @@ def setup(request, session):
         def test_something(setup, session):
             assert session.url == "https://example.org"
 
-    The session is closed when the test module goes out of scope by an
-    implicit call to `session.end`.
+    When the test function goes out of scope, any remaining user prompts
+    and opened windows are closed, and the current browsing context is
+    switched back to the top-level browsing context.
+
+    After returning the session to a near-pristine state, the session
+    is also closed by an implicit call to `session.end`.
     """
 
     def __init__(self, client):
         self.client = client
 
-    @pytest.fixture(scope="module")
+        # store handle of original window
+        # so that we do not close it
+        self.original_window = self.client.window_handle
+        self.client.end()
```

Comment:
Excellent question and one I’d be happy to explain!

Firstly, `client` refers to the WebDriver client’s [_Session_](https://github.com/w3c/wdclient/blob/master/webdriver/client.py#L137) instance. I chose to split this out as a separate library so that both the tests in WPT and the harness in wptrunner could have access to the symbols (for imports, and so on).

In order to test session creation, the client does not start a session for you on construction. You have to explicitly call [_webdriver.client.Session.start_](https://github.com/w3c/wdclient/blob/master/webdriver/client.py#L152). However, when you call a regular command (i.e. not `Session.start` or `Session.end`), new sessions are created _implicitly_ in https://github.com/w3c/wdclient/blob/master/webdriver/client.py#L13.

This makes `self.client.window_handle` implicitly start a session in the code above, but in order to test session creation in the tests we end it right afterwards so that the tests are run in a sessionless state by default.